File: src/code_review_agent/agents/exceptions.py
"""Shared exception types for the review-stage and lead-engineer agents."""

from httpx import TransportError
from strands.types.exceptions import (
    EventLoopException,
    MCPClientInitializationError,
    ToolProviderException,
)

INFRA_EXCEPTIONS: tuple[type[BaseException], ...] = (
    EventLoopException,
    MCPClientInitializationError,
    ToolProviderException,
    TransportError,
)
"""Exceptions treated as infrastructure failures rather than isolated/business
errors (model connection loss, GitHub MCP client init failure, transport-level
timeouts). Callers should re-raise these instead of degrading them to a
business-level error/empty result, so the A2A task boundary handler
(``except Exception`` in ``api/agents/*.py``) marks the task as failed instead
of silently completing with partial data.
"""


class StructuredOutputMissingError(RuntimeError):
    """Raised when an LLM agent call ends without a structured output result.

    Strands does not raise when a turn/token limit is exhausted: it returns an
    ``AgentResult`` with ``stop_reason`` set (e.g. ``"limit_turns"``) and
    ``structured_output=None`` instead. Callers must check for this explicitly
    rather than assuming ``result.structured_output`` is always populated —
    otherwise the failure surfaces later as an opaque ``AttributeError`` on
    whatever field is accessed first.
    """

    def __init__(self, agent_label: str, stop_reason: str | None) -> None:
        super().__init__(
            f"{agent_label} completed without producing structured output "
            f"(stop_reason={stop_reason!r}). The model likely could not satisfy "
            "the output schema within the configured turn limit."
        )

File: packages/agent-core/src/agents/exceptions.ts
import {
  ContextWindowOverflowError,
  MaxTokensError,
  ModelError,
  ModelThrottledError,
  type StopReason,
} from "@strands-agents/sdk";
import { GithubMcpConnectionError } from "../tools/github-mcp.js";

/**
 * Raised when an LLM agent call ends without a structured output result.
 *
 * The SDK does not throw when a turn/token limit is exhausted: `Agent.invoke()`
 * resolves with an `AgentResult` whose `stopReason` is set (e.g. `"limitTurns"`)
 * and `structuredOutput` is `undefined`. Callers must check for this explicitly
 * rather than assuming `result.structuredOutput` is always populated.
 */
export class StructuredOutputMissingError extends Error {
  constructor(agentLabel: string, stopReason: StopReason | undefined) {
    super(
      `${agentLabel} completed without producing structured output ` +
        `(stopReason=${String(stopReason)}). The model likely could not satisfy ` +
        "the output schema within the configured turn limit.",
    );
    this.name = "StructuredOutputMissingError";
  }
}

/**
 * Classifies an error as an infrastructure failure (model connection loss,
 * GitHub MCP client init failure) rather than a business/content-level error.
 *
 * Mirrors Python's `INFRA_EXCEPTIONS` tuple: the review orchestrator (slice C)
 * re-throws infra failures instead of degrading them to an isolated
 * `ReviewError`, so the whole batch aborts rather than silently completing
 * with partial data. `ContextWindowOverflowError`/`MaxTokensError`/
 * `ModelThrottledError` are excluded even though they extend `ModelError`:
 * they are content/rate-limit issues scoped to one reviewer's call, not a
 * system-wide connectivity problem, matching how Python's `INFRA_EXCEPTIONS`
 * never included content-shape failures either. See
 * typescript-agents-tools-migration-spec.md section 5.2.
 */
export function isInfraError(error: unknown): boolean {
  if (error instanceof GithubMcpConnectionError) {
    return true;
  }
  if (
    error instanceof ContextWindowOverflowError ||
    error instanceof MaxTokensError ||
    error instanceof ModelThrottledError
  ) {
    return false;
  }
  return error instanceof ModelError;
}
